File: 백준/Silver/2178. 미로 탐색/미로 탐색.swift
import Foundation

let input = readLine()!.split(separator: " ").map{Int($0)!}
let (n,m) = (input[0], input[1])
let dir = [
    (1,0), (-1,0), (0,1),(0,-1)
]

var board = [[Int]]()
for _ in 0..<n {
    let row = readLine()!.map { Int(String($0))! }
    board.append(row)
}
var visited = Array(repeating: Array(repeating: false, count: m), count: n)
var cnt = Array(repeating: Array(repeating: 0, count: m), count: n)

func bfs(_ r: Int,_ c: Int) {
    var q = [(r,c)]
    visited[r][c] = true
    cnt[r][c] = 1
    var idx = 0
    while idx < q.count {
        let (nr, nc) = q[idx]
        idx += 1
        for (dx, dy) in dir {
            var sr = nr + dx
            var sc = nc + dy
            if sr >= 0 && sr < n && sc >= 0 && sc < m && board[sr][sc] == 1 && !visited[sr][sc] {
                q.append((sr,sc))
                visited[sr][sc] = true
                cnt[sr][sc] = cnt[nr][nc] + 1
            }
        }
    }
}

bfs(0,0)
print(cnt[n-1][m-1])

File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
import sys
from collections import deque

input = sys.stdin.readline
n,m = map(int, input().split())
miro = [list(map(int, input().strip())) for _ in range(n)]
visited = [[False] * m for _ in range(n)]
dirs = [(1,0),(-1,0),(0,1),(0,-1)]
distance = [[0]*m for _ in range(n)]

def bfs(r,c):
    q = deque([(r,c)])
    visited[r][c] = True
    while q:
        (nr, nc) = q.popleft()
        for dx, dy in dirs:
            x,y = nr+dx, nc+dy
            if 0<=x<n and 0<=y<m and miro[x][y] == 1 and not visited[x][y]:
                distance[x][y] = distance[nr][nc]+1
                q.append((x,y))
                visited[x][y] = True

bfs(0,0)
print(distance[n-1][m-1]+1)